R/plot.data_table_threads_benchmark.R

```diff
@@ -16,46 +16,56 @@
 {
-  df <- x
-  rownames(df) <- NULL
-  df$speedup <- df$medianTime[df$threadCount == 1] / df$medianTime
-
-  setDT(df)
-  maxSpeedup <- df[, .(threadCount = threadCount[which.max(speedup)], speedup = max(speedup)), by = expr]
-  idealSpeedup <- seq(1, getDTthreads())
-  idealSpeedupData <- data.frame(threadCount = 1:getDTthreads(), speedup = idealSpeedup)
-  subOptimalSpeedupData <- data.frame(threadCount = seq(1, getDTthreads(), length.out = getDTthreads()), speedup = seq(1, getDTthreads()/2, length.out = getDTthreads()))
-
-  closestPoints <- data.frame()
-  for(i in unique(df$expr)) {
-    dfSubset <- df[df$expr == i, ]
-    suboptimalSubset <- subOptimalSpeedupData[subOptimalSpeedupData$threadCount %in% dfSubset$threadCount, ]
-    closestPoint <- dfSubset[which.max(dfSubset$speedup - suboptimalSubset$speedup), ]
-    closestPoints <- rbind(closestPoints, closestPoint)
+  x[, `:=`(speedup = medianTime[threadCount == 1] / medianTime, type = "Measured"), by = expr]
+
+  systemThreadCount <- getDTthreads()
+  functions <- unique(x$expr)
+
+  speedupData <- data.table(
+    expr = rep(functions, each = systemThreadCount),
+    threadCount = rep(c(1:systemThreadCount, seq(1, systemThreadCount, length.out = systemThreadCount)), length(functions)),
+    speedup = c(rep(seq(1, systemThreadCount), length(functions)), rep(seq(1, systemThreadCount / 2, length.out = systemThreadCount), length(functions))),
+    type = rep(c("Ideal", "Recommended"), each = systemThreadCount * length(functions))
+  )
+
+  for(col in setdiff(names(x), names(speedupData))) 
+  {
+    speedupData[, (col) := NA]
   }
 
-  ggplot(df, aes(x = threadCount, y = speedup)) +
-    geom_line(aes(linetype = "Measured")) +
-    geom_line(data = idealSpeedupData, aes(x = threadCount, y = speedup, linetype = "Ideal"), color = "black") +
-    geom_line(data = subOptimalSpeedupData, aes(x = threadCount, y = speedup, linetype = "Sub-optimal"), color = "black") +
-    geom_point(data = closestPoints, aes(x = threadCount, y = speedup, shape = "Recommended"), color = "black", size = 2) +
-    geom_point(data = maxSpeedup, aes(x = threadCount, y = speedup, shape = "Best performing"), color = "red", size = 2) +
-    geom_text(data = closestPoints, aes(label = threadCount), vjust = -0.5, size = 4, na.rm = TRUE) +
-    geom_text(data = maxSpeedup, aes(label = threadCount), vjust = -0.5, size = 4, na.rm = TRUE) +
-    geom_ribbon(aes(ymin = speedup - 0.3, ymax = speedup + 0.3), alpha = 0.5) +
+  maxSpeedup <- x[, .(threadCount = threadCount[which.max(speedup)], speedup = max(speedup), type = "Ideal"), by = expr]
+
+  combinedLineData <- rbind(speedupData, x, use.names = TRUE, fill = TRUE)
+
+  closestPoints <- x[, {
+    recommendedSpeedupSubset <- speedupData[expr == .BY$expr & type == "Recommended"]
+    merged <- .SD[recommendedSpeedupSubset, on = .(threadCount), nomatch = 0L]
+    .SD[which.max(speedup - merged$speedup)]
+  }, by = expr]
+
+  closestPoints[, `:=`(type = "Recommended")]
```

Comment:
rather than repeating closestPoints here you could  add the type column inside, `data.table(type="Recommended", .SD[which.max(...)])` or chain 
```r
}, by = expr
][
, type := "Recommended"
]
```
Also `` `:=`(type="Recommended", foo="bar") `` should only be used (backticks version of :=) when there are multiple columns to assign at the same time.